### PR TITLE
add patchJson6902 transformer

### PR DESCRIPTION
**Pull request by Liujingfang1**

### CODE REVIEW

**Comment by Liujingfang1** [comment]
There are two things that we may do it differently.
- Move this transformer to `pkg/patch` lib
- Remove the dependency on `loader`, use `map[ResId][]byte` and this map is constructed in `application` level as other resources

File: pkg/transformers/patch/patchtransformer.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+// patchTransformer applies different types of patches.
+type patchTransformer struct {
+	transformers []transformers.Transformer
+}
+
+var _ transformers.Transformer = &patchTransformer{}
+
+// NewPatchTransformer constructs a patchTransformer.
+func NewPatchTransformer(slice []*resource.Resource, patchesJ6 map[resource.ResId][]byte) (transformers.Transformer, error) {
```
Comment (mengqiy) [inline pkg/transformers/patch/patchtransformer.go]: > patchesJ6

This var is strange. You can use `patchesJson` or `jsonPatches`.

File: pkg/app/application.go [hunk 2]
```diff
@@ -37,6 +36,7 @@ import (
 	"github.com/kubernetes-sigs/kustomize/pkg/resmap"
 	"github.com/kubernetes-sigs/kustomize/pkg/resource"
 	"github.com/kubernetes-sigs/kustomize/pkg/transformers"
+	patchtransformer "github.com/kubernetes-sigs/kustomize/pkg/transformers/patch"
```
Comment (monopole) [inline pkg/app/application.go]: The code down here should be in pkg/patch, so this wouldn't be necessary.

File: pkg/transformers/patch/patchtransformer.go (hunk 1 above)
Comment (mengqiy) [inline pkg/transformers/patch/patchtransformer.go]: IIUC `slice` here is actually SMPatches.
So we may want to change it to a more descriptive name.

File: pkg/transformers/patch/patchtransformer.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+// Transform apply the patches on top of the base resources.
+func (pt *patchTransformer) Transform(baseResourceMap resmap.ResMap) error {
+	return pt.transformWithCheckConflicts(baseResourceMap)
+}
+
+func (pt *patchTransformer) transformWithCheckConflicts(m resmap.ResMap) error {
+	mcopy := resmap.ResMap{}
+	for id, res := range m {
```
Comment (mengqiy) [inline pkg/transformers/patch/patchtransformer.go]: Does it make sense to pull this out as a separate method of `resmap.ResMap`?
Maybe call it `func (rm *ResMap) Deepcopy() *ResMap`?

File: pkg/transformers/patch/patchtransformer_test.go [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+							},
+						},
+					},
+				},
+			},
+		},
+		),
+	}
+
+	patchJ6 := map[resource.ResId][]byte{
```
Comment (mengqiy) [inline pkg/transformers/patch/patchtransformer_test.go]: > patchJ6

I don't like this var name

File: pkg/app/application.go [hunk 5]
```diff
@@ -160,12 +157,19 @@ func (a *Application) loadCustomizedResMap() (resmap.ResMap, error) {
 		errs.Append(errors.Wrap(err, "NewResourceSliceFromPatches"))
 	}
 
+	if len(errs.Get()) > 0 {
+		return nil, errs
+	}
+	patchesJson6902, err := patch.NewPatchJson6902(a.ldr, a.kustomization.PatchesJson6902)
```
Comment (monopole) [inline pkg/app/application.go]: Just `p` - identifiers have a length proportional to their scope.

File: pkg/patch/loadpatch.go [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+package patch
+
+import (
+	"github.com/kubernetes-sigs/kustomize/pkg/loader"
+	"github.com/kubernetes-sigs/kustomize/pkg/resource"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// NewPatchJson6902 load a slice of PatchJson6902
+func NewPatchJson6902(l loader.Loader, slice []PatchJson6902) (map[resource.ResId][]byte, error) {
```
Comment (monopole) [inline pkg/patch/loadpatch.go]: Why can't this return a resmap?

File: pkg/transformers/patch/patchtransformer_test.go [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+			},
+			"spec": map[string]interface{}{
+				"template": map[string]interface{}{
+					"spec": map[string]interface{}{
+						"containers": []interface{}{
+							map[string]interface{}{
+								"name": "nginx",
+								"env": []interface{}{
+									map[string]interface{}{
+										"name":  "ANOTHERENV",
```
Comment (mengqiy) [inline pkg/transformers/patch/patchtransformer_test.go]: You will need to drop these 2 lines to test when there are conflicts between SMP and JSON Patch.
If 2 SMPs already have conflicts, you may not even be able to reach the functionality that handles conflicts between SMP and JSON Patch.

File: pkg/app/application.go [hunk 8]
```diff
@@ -160,12 +157,19 @@ func (a *Application) loadCustomizedResMap() (resmap.ResMap, error) {
 		errs.Append(errors.Wrap(err, "NewResourceSliceFromPatches"))
```
Comment (monopole) [inline pkg/app/application.go]: Please move L154 into a function called, say, `handleAliasedAndDeprecatedFields`, and call it at the top.

File: pkg/patch/loadpatch.go [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+			p.Target.Name,
+			"",
+			p.Target.Namespace,
+		)
+		content, err := l.Load(p.Path)
+		if err != nil {
+			return nil, err
+		}
+		if val, ok := patches[id]; ok {
+			content = append(val, content...)
```
[the remark this answers is not in the document]
Comment (Liujingfang1) [reply pkg/patch/loadpatch.go]: This is to append at the end. `val` is the aggregated JSON patch found so far. We append a new JSON patch `content` to it.

File: pkg/transformers/patch/patchtransformer.go (hunk 1 above)
Comment (Liujingfang1) [reply pkg/transformers/patch/patchtransformer.go]: yep, I will change it.

Comment (Liujingfang1) [reply pkg/transformers/patch/patchtransformer.go]: sounds good.

File: pkg/transformers/patch/patchtransformer.go (hunk 3 above)
Comment (Liujingfang1) [reply pkg/transformers/patch/patchtransformer.go]: Yes, it makes sense to put it as a method.

File: pkg/transformers/patch/patchtransformer_test.go (hunk 7 above)
Comment (Liujingfang1) [reply pkg/transformers/patch/patchtransformer_test.go]: Good catch. I copied this test from the previous ones and missed that part.

File: pkg/transformers/patch/patchtransformer_test.go (hunk 4 above)
Comment (Liujingfang1) [reply pkg/transformers/patch/patchtransformer_test.go]: Haha, I was lazy on those var names.

**Pull request closed without merging**